win32/pe-parser.lua

```diff
@@ -525,44 +525,39 @@ function M.msvcrt(infile)
   end
   local obj, err = M.parse(path..file)
   if not obj then return obj, err end
-  
+
   for i, dll in ipairs(obj.DataDirectory.ImportTable) do
     dll = dll.Name:upper()
-	  local result = dll:match('(MSVCR%d*D?)%.DLL')
-	  if not result then
-	    result = dll:match('(MSVCRTD?)%.DLL')
-	  end
-	  if not result then
-	    result = dll:match('(VCRUNTIME%d*D?)%.DLL')
-	  end
+    local result = dll:match('(MSVCR%d*D?)%.DLL')
+    if not result then
+      result = dll:match('(MSVCRTD?)%.DLL')
+    end
+    if not result then
+      result = dll:match('(VCRUNTIME%d*D?)%.DLL')
+    end
+    if not result then
+      result = dll:match('(UCRTBASED?)%.DLL')
+    end
+    if not result then
+      -- api-ms-win-crt-xxx also indicate the universal runtime
+      result = dll:match('(API%-MS%-WIN%-CRT%-RUNTIME%-)')
+      if result then
+        result = "UCRTBASE"
+      end
+    end
     -- success, found it return name + binary where it was found
     if result then return result, infile end
   end
-  
+
   -- not found, so traverse all imported dll's
   for i, dll in ipairs(obj.DataDirectory.ImportTable) do
     local rt, ref = M.msvcrt(path..dll.Name)
-    if rt then 
+    if rt then
       return rt, ref  -- found it
     end
   end
 
   return nil, "No msvcrt found"
 end
 
-function M.get_architecture(program)
```

Comment:
is this correct to remove? was this maybe a luarocks specific function that was added? not sure though

Reply:
Hm, this was added in d90c3c918a2555eff8776e1acabab7a777f13c31, when pe-parser was in `src/luarocks/fs/`. There are no references currently to this function anywhere in the repository so maybe it's fine to remove it, but I'm struggling to figure out why it was added in the first place...

Reply:
This was moved there because it was part of `install.bat` as part of an effort to move logic out of `install.bat` and into LuaRocks modules proper.

I still think it would be a good idea to move it into pe-parser, essentially as a high-level utility API (note that that version takes a `program` as an argument, versus the one in `install.bat` which hardcodes `vars.LUA`)

Reply:
I don't think that'll work. PE-parser recognizes far more formats, hence the simple implementation for detecting 32/64 bits doesn't suffice. It might for LuaRocks, but not for the parser itself.

Here's the list of archs:
https://github.com/Tieske/pe-parser/blob/acbcd85a337dcbfbda2c7511b53c56666337d3f8/src/pe-parser.lua#L29-L52

As per ChatGPT:

Here are the common names for each of the machine architectures listed in your PortableExecutable introspection data:

- `IMAGE_FILE_MACHINE_UNKNOWN` (0): Unknown machine type
- `IMAGE_FILE_MACHINE_AM33` (1d3): Matsushita AM33
- `IMAGE_FILE_MACHINE_AMD64` (8664): x64 (AMD or Intel)
- `IMAGE_FILE_MACHINE_ARM` (1c0): ARM little endian
- `IMAGE_FILE_MACHINE_ARMNT` (1c4): ARM Thumb-2 little endian
- `IMAGE_FILE_MACHINE_ARM64` (aa64): ARM64 little endian
- `IMAGE_FILE_MACHINE_EBC` (ebc): EFI byte code
- `IMAGE_FILE_MACHINE_I386` (14c): Intel 386 or later processors and compatible processors
- `IMAGE_FILE_MACHINE_IA64` (200): Intel Itanium processor family
- `IMAGE_FILE_MACHINE_M32R` (9041): Mitsubishi M32R little endian
- `IMAGE_FILE_MACHINE_MIPS16` (266): MIPS16
- `IMAGE_FILE_MACHINE_MIPSFPU` (366): MIPS with FPU
- `IMAGE_FILE_MACHINE_MIPSFPU16` (466): MIPS16 with FPU
- `IMAGE_FILE_MACHINE_POWERPC` (1f0): Power PC little endian
- `IMAGE_FILE_MACHINE_POWERPCFP` (1f1): Power PC with floating point support
- `IMAGE_FILE_MACHINE_R4000` (166): MIPS little endian
- `IMAGE_FILE_MACHINE_SH3` (1a2): Hitachi SH3
- `IMAGE_FILE_MACHINE_SH3DSP` (1a3): Hitachi SH3 DSP
- `IMAGE_FILE_MACHINE_SH4` (1a6): Hitachi SH4
- `IMAGE_FILE_MACHINE_SH5` (1a8): Hitachi SH5
- `IMAGE_FILE_MACHINE_THUMB` (1c2): ARM or Thumb ("thumb")
- `IMAGE_FILE_MACHINE_WCEMIPSV2` (169): MIPS little-endian WCE v2

These names are more commonly used and recognizable when referring to these specific machine architectures in the context of Portable Executable files.

The PR only assumes `IMAGE_FILE_MACHINE_I386` to be `x86` (32 bit), and everything else to be `x86_64` (64 bit). And hence misses the ARM architectures, as well Itanium.

I do not see an easy way to add this to pe-parser. So imho this should probably stay within LuaRocks in a simple wrapper.

Reply:
> As per ChatGPT:

Please do not use ChatGPT as a source of technical information in any project I'm involved with. It constantly produces false data, and having to double-check everything it produces is equivalent to just skipping it altogether and going to a reliable source anyway.

This link illustrates another occasion where this has happened to me before, and the "reference" info provided by ChatGPT was a blatant lie:
https://mastodon.social/@hanshuebner/109952292956659419

Reply:
> The PR only assumes IMAGE_FILE_MACHINE_I386 to be x86 (32 bit), and everything else to be x86_64 (64 bit). And hence misses the ARM architectures, as well Itanium.
> 
> I do not see an easy way to add this to pe-parser. So imho this should probably stay within LuaRocks in a simple wrapper.

Fair point! We can still move it out of `install.bat` in a separate PR, and not mix it with the vendored code of pe-parser.

Reply:
Turns out that sysdetect (also vendored-in) already does something similar (covering at least the set of architectures our users seem to care about)

https://github.com/luarocks/sysdetect/blob/main/src/sysdetect.tl#L401-L408

Reply:
> Please do not use ChatGPT as a source of technical information in any project I'm involved with. It constantly produces false data, and having to double-check everything it produces is equivalent to just skipping it altogether and going to a reliable source anyway.

I totally agree (and I know your stance), hence I specifically mentioned it as chatgpt generated. The only aim here was to list the more common names of the platforms, to proof the point. It wasn't about being technically accurate.